DOCSP-41762 Add transactions page (#569)
mongodb/docs-java: source/fundamentals/transactions.txt

Overview
--------

In this guide, you can learn how to use the {+driver-short+} to perform 
**transactions**. :manual:`Transactions </core/transactions/>` allow
you to run a series of operations that do not change any data until the
transaction is committed. If any operation in the transaction returns an
error, the driver cancels the transaction and discards all data changes
before they ever become visible.

In MongoDB, transactions run within logical **sessions**. A
:manual:`session </reference/server-sessions/>` is a grouping of related
read or write operations that you intend to run sequentially. Sessions
enable :manual:`causal consistency
</core/read-isolation-consistency-recency/#causal-consistency>` for a
group of operations or allow you to execute operations in an
:website:`ACID transaction </basics/acid-transactions>`. MongoDB
guarantees that the data involved in your transaction operations remains
consistent, even if the operations encounter unexpected errors.

When using the {+driver-short+}, you can create a new session from a 
``Client`` instance as a ``ClientSession``. We recommend that you reuse
your client for multiple sessions and transactions instead of

> instantiating a new client each time.

Reviewer: I: double check all of the type references on this page to make sure theyre the correct ones. Pretty sure that the client type is called `MongoClient` in the Jvm drivers

   Use a ``ClientSession`` only with the ``MongoClient`` (or associated
   ``MongoDatabase`` or ``MongoCollection``) that created it. Using a
   ``ClientSession`` with a different ``MongoClient`` results in operation
   errors.

Methods
-------

Create a ``ClientSession`` by using the ``startSession()`` method on your 

> ``MongoClient``instance. You can then modify the session state by using the 

Reviewer: I; causing a rendering error
```suggestion
``MongoClient`` instance. You can then modify the session state by using the 
```

> methods provided by the ``ClientSession``. The following table details the 
> methods you can use to manage your transaction:

Reviewer: ```suggestion
methods provided by the ``ClientSession``. The following table describes the 
methods you can use to manage your transaction:
```

.. list-table::
   :widths: 25 75
   :header-rows: 1

   * - Method
     - Description

   * - ``startTransaction()``

>      - | Starts a new transaction for this session with the
>          default transaction options. Use the ``TransactionOptions`` 
>          parameter to start a transaction with given options. You 
>          cannot start a transaction if there's already an active transaction
>          on the session.

Reviewer: ```suggestion
     - | Starts a new transaction for this session with the
         default transaction options. Pass an instance of ``TransactionOptions`` 
         as a parameter to start a transaction with specified options. You 
         cannot start a transaction if there's already an active transaction
         running in the session.
```

       |

>        | **Parameter**: ``transactionOptions``

Reviewer: ```suggestion
       | **Parameter**: ``TransactionOptions transactionOptions``
```

   * - ``abortTransaction()``
     - | Ends the active transaction for this session. Returns an error 
         if there is no active transaction for the
         session or the transaction was previously ended.

   * - ``commitTransaction()``
     - | Commits the active transaction for this session. Returns an
         error if there is no active transaction for the session or if the
         transaction was ended.

   * - ``withTransaction()``

>      - | Starts a new transaction for this session and runs the given function.

Reviewer: S: add to this description some info about defining a transactionBody

>      - | Starts a new transaction for this session and runs the given function.

Reviewer: S: in this description, you can also mention that this method comprises the "Convenient Transaction API" because it starts commits and handles errors all in the same method.

       |

>        | **Parameter**: ``transactionBody``

Reviewer: ```suggestion
       | **Parameter**: ``TransactionBody<T> transactionBody``
```

A ``ClientSession`` also has methods to retrieve session properties and modify mutable 
session properties. View the `API documentation <{+api+}/apidocs/mongodb-driver-sync/com/mongodb/client/ClientSession.html>`__ 
to learn more about these methods.

> 

Reviewer: S: instead of linking to clientsession api docs, just reroute to the api docs section at the bottom of this guide

Author: QQ what's the syntax for this in rst? I didn't see this in the slide deck and what I found in other references didn't seem to work. Still getting used to using this instead of markdown!

Author: Think I figured it out actually!

Example
-------

The following example demonstrates how you can create a session, create a transaction, 
and commit a multi-document insert operation through the following steps:

> 1. Create a session from the client using the ``startSession()`` method.
> #. Set transaction options to configure transactional behavior.
> #. Use the ``withTransaction()`` method to start a transaction.
> #. Insert multiple documents. The ``withTransaction()`` method executes, commits, and aborts 
>    the transaction. If any operation results in errors, ``withTransaction()`` handles canceling
>    the transaction.

Reviewer: ```suggestion
1. Create a session from the client by using the ``startSession()`` method.
#. Set transaction options to configure transaction behavior.
#. Use the ``withTransaction()`` method to start a transaction.
#. Insert multiple documents. The ``withTransaction()`` method executes, commits, and aborts 
   the transaction. If any operation results in errors, ``withTransaction()`` handles canceling
   the transaction.
```

.. literalinclude:: /includes/fundamentals/code-snippets/Transaction.java
   :language: java
   :start-after: start transaction

>    :end-before: end transaction

Reviewer: S: add the :dedent: option to this block to remove the indentation

If you require more control over your transactions, you can use the ``startTransaction()``
method. This method allows you to explicitly commit or abort the transaction and manually

> manage the transaction lifecycle.

Reviewer: S: if you use the startTxn() method, you also need to mention that the user will need to use the other methods to commit/handle errors
